### PR TITLE
Feature/pipeline misconfigurations scan

### PR DESCRIPTION
**Pull request by tamirkiviti13**

### CODE REVIEW

File: pkg/pipelines/rules/policies/pipeline_scanned_for_secrets.rego [hunk 1]
```diff
@@ -0,0 +1,96 @@
+package builtin.pipeline.SECRET_SCANNING
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "SECRET_SCANNING",
+	"avd_id": "",
```
Comment (tzurielweisberg) [inline pkg/pipelines/rules/policies/pipeline_scanned_for_secrets.rego:7]: Dont you want to have a dedicated id? for the DB?

**Review by tzurielweisberg — APPROVED** [review]

File: pkg/pipelines/pipelines.go [hunk 2]
```diff
@@ -93,25 +101,74 @@ func enhancePipelines(rootDir string, pipelines []*buildsecurity.Pipeline) error
...
+	}
+	files = append(files, gitlabFiles...)
+
+	if err != nil {
+		return nil, nil, err
+	}
 
 	if err := enhancePipelines(rootDir, pipelines); err != nil {
-		return nil, err
+		fmt.Println("Failed to enhance pipelines:", err)
```
Comment (orizerah) [inline pkg/pipelines/pipelines.go:171]: Use logger

File: pkg/pipelines/rules/lib/pipeline.rego [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		`console\.assert\(`,
+		`process\.stdout\.write\(`,
+		`echo`,
+		`print\(`,
+		`logging\.info\(`,
+		`logging\.warning\(`,
+		`logging\.error\(`,
+		`logging\.critical\(`,
+		`logging\.exception\(`,
+		`logging\.log\(`,
```
Comment (orizerah) [inline pkg/pipelines/rules/lib/pipeline.rego:73]: Unify all of the `logging.*` regexes

File: pkg/pipelines/rules/policies/checkut_with_pr_target.rego [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package builtin.pipeline.CHECKOUT_WITH_PR_TARGET
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/checkut_with_pr_target.rego:1]: Filename typo

File: pkg/scanner/scanner.go [hunk 5]
```diff
@@ -108,3 +130,221 @@ func MatchTriggeredBy(triggeredBy string) buildsecurity.TriggeredByEnum {
...
+
+	scanResults, err := pipelineScanner.ScanFS(context.WithValue(ctx, "sourcesMap", sourcesMap), memFs, ".")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed scan pipelines")
+	}
+	results := misconfsToResults(resultsToMisconf("pipeline", pipelineScanner.Name(), scanResults))
+	return results, nil
+}
+
+func getPolicyReaders() ([]io.Reader, error) {
```
Comment (orizerah) [inline pkg/scanner/scanner.go:174]: Comment

File: pkg/pipelines/scanner.go [hunk 6]
```diff
@@ -101,6 +103,9 @@ func (s *Scanner) ScanFS(ctx context.Context, fs fs.FS, path string) (scan.Resul
 		})
 	}
 
+	k, _ := json.MarshalIndent(inputs, "", "  ")
```
Comment (orizerah) [inline pkg/pipelines/scanner.go:106]: Is this on purpose?

File: pkg/pipelines/rules/policies/untrusted_input_usage.rego [hunk 7]
```diff
@@ -25,7 +25,7 @@ deny[result] {
 	count(path) == 1
 	pipeline.contains_untrusted_inputs(value)
 	result := {
-		"msg": value,
+		"msg": sprintf("consider evaluating user inputs in intermediate environment variables instead of using them right in '%s' step shell", [input.jobs[i].steps[j]]),
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/untrusted_input_usage.rego:28]: This one starts with a lower case letter, the one above starts with a capital letter.

File: pkg/pipelines/rules/policies/insecure_fetching.rego [hunk 8]
```diff
@@ -4,7 +4,7 @@ import data.lib.pipeline
 
 __rego_metadata__ := {
 	"id": "INSECURE_FETCHING",
-	"avd_id": "",
+	"avd_id": "AVD-PIPELINE-0009",
 	"title": "insecure fetching",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/insecure_fetching.rego:8]: Capital letter title

File: pkg/pipelines/rules/policies/variable_logging.rego [hunk 9]
```diff
@@ -4,7 +4,7 @@ import data.lib.pipeline
 
 __rego_metadata__ := {
 	"id": "VARIABLES_LOGGING",
-	"avd_id": "",
+	"avd_id": "AVD-PIPELINE-0003",
 	"title": "echo of variables",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/variable_logging.rego:8]: Capitalize title

File: pkg/pipelines/rules/policies/extra_index_url.rego [hunk 10]
```diff
@@ -0,0 +1,32 @@
+package builtin.pipeline.EXTRA_INDEX_URL
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "EXTRA_INDEX_URL",
+	"avd_id": "AVD-PIPELINE-0001",
+	"title": " using extra-index-url",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/extra_index_url.rego:8]: Capitalize title

File: pkg/pipelines/rules/policies/eval_command.rego [hunk 11]
```diff
@@ -4,7 +4,7 @@ import data.lib.pipeline
 
 __rego_metadata__ := {
 	"id": "EVAL_COMMAND",
-	"avd_id": "",
+	"avd_id": "AVD-PIPELINE-0004",
 	"title": "eval command",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/eval_command.rego:8]: Capitalize title

File: pkg/pipelines/rules/policies/checkout_with_pr_target.rego [hunk 12]
```diff
@@ -4,7 +4,7 @@ import data.lib.pipeline
 
 __rego_metadata__ := {
 	"id": "CHECKOUT_WITH_PR_TARGET",
-	"avd_id": "",
+	"avd_id": "AVD-PIPELINE-0002",
 	"title": "checkout with pull_request_target",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/checkout_with_pr_target.rego:8]: Capitalize

File: pkg/pipelines/rules/policies/checkout_persist_credentials.rego [hunk 13]
```diff
@@ -4,7 +4,7 @@ import data.lib.pipeline
 
 __rego_metadata__ := {
 	"id": "PERSIST_CREDENTIALS",
-	"avd_id": "AVD-ID-2",
+	"avd_id": "AVD-PIPELINE-0005",
 	"title": "persist-credentials is true",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/checkout_persist_credentials.rego:8]: Capitalize

File: pkg/pipelines/rules/policies/eval_command.rego [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "EVAL_COMMAND",
+	"avd_id": "AVD-PIPELINE-0004",
+	"title": "eval command",
+	"severity": "LOW",
+	"type": "Pipeline Yaml Security Check",
+	"description": "Evaluate command in pipeline(s)",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/eval_command.rego:11]: Add a `.` in the end of the description

File: pkg/pipelines/rules/policies/insecure_fetching.rego [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "INSECURE_FETCHING",
+	"avd_id": "AVD-PIPELINE-0009",
+	"title": "insecure fetching",
+	"severity": "MEDIUM",
+	"type": "Pipeline Yaml Security Check",
+	"description": "Unencrypted fetching request",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/insecure_fetching.rego:11]: `.`

File: pkg/pipelines/rules/policies/pipeline_scanned_for_vulns.rego [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "VULN_SCANNING",
+	"avd_id": "AVD-PIPELINE-0021",
+	"title": "Ensure pipelines are automatically scanned for vulnerabilities",
+	"severity": "HIGH",
+	"type": "Pipeline Yaml Security Check",
+	"description": "Scan pipelines for vulnerabilities. It is recommended to do that automatically",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/pipeline_scanned_for_vulns.rego:11]: `.`

File: pkg/pipelines/rules/policies/variable_logging.rego [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+import data.lib.pipeline
+
+__rego_metadata__ := {
+	"id": "VARIABLES_LOGGING",
+	"avd_id": "AVD-PIPELINE-0003",
+	"title": "echo of variables",
+	"severity": "LOW",
+	"type": "Pipeline Yaml Security Check",
+	"description": "Avoid printing variables. These contain sensitive data, secrets, credentials, and keys. printing them would result in them exposed and printed in plain text. In addition, avoid printing untrusted input that can be manipulated by attacker",
```
Comment (orizerah) [inline pkg/pipelines/rules/policies/variable_logging.rego:11]: `.`

**Review by orizerah — APPROVED** [review]

**Pull request merged**